File: kmp.py
#!/usr/bin/env python3
# kmp.py - KMP program by Sergey 2016
# AlgoArt - The Art of Algorithms (github.com/algoart/algoart)

"""

In computer science, the Knuth-Morris-Pratt string searching algorithm
(or KMP algorithm) searches for occurrences of a "word" W within a main
"text string" S by employing the observation that when a mismatch occurs, the
word itself embodies sufficient information to determine where the next match
could begin, thus bypassing re-examination of previously matched characters.

A string matching algorithm wants to find the starting index m in string S[n]
that matches the search word W[k]

Time complexity:    O(n+k)
Extra space:        O(k)

"""

import unittest
import sys

###############################################################################
# KMP (Main Program)
###############################################################################


# KMP partial match table
def kmp_table(W):
    assert(len(W) > 0)
    cnd = -1
    T = [-1]
    for i in range(len(W)):
        while True:
            if cnd == -1 or W[cnd] == W[i]:
                cnd += 1
                T.append(cnd)
                break
            cnd = T[cnd]
    return T


# KMP search
def kmp_search(W, S):
    cnd = 0
    T = kmp_table(W)
    for i in range(len(S)):
        while True:
            if cnd == -1 or W[cnd] == S[i]:
                cnd += 1
                break
            cnd = T[cnd]
        if cnd == len(W):
            return i - cnd + 1
    return -1


###############################################################################
# Unit Tests
###############################################################################


class unitTests(unittest.TestCase):

    def test_Kmp_(self):
        """ Gcd class testing """

        # KMP table generation
        self.assertEqual(kmp_table("A"), [-1, 0])
        self.assertEqual(kmp_table("AA"), [-1, 0, 1])
        self.assertEqual(kmp_table("AB"), [-1, 0, 0])
        self.assertEqual(kmp_table("ABAB"), [-1, 0, 0, 1, 2])
        self.assertEqual(kmp_table("ABCDABA"), [-1, 0, 0, 0, 0, 1, 2, 1])

        # KMP search
        self.assertEqual(kmp_search("A", ""), -1)
        self.assertEqual(kmp_search("AB", "B"), -1)
        self.assertEqual(kmp_search("A", "BBBB"), -1)
        self.assertEqual(kmp_search("A", "A"), 0)
        self.assertEqual(kmp_search("BA", "BA"), 0)
        self.assertEqual(kmp_search("A", "BA"), 1)
        self.assertEqual(kmp_search("AS", "BDAS"), 2)
        self.assertEqual(kmp_search("ABAC", "ABABABACB"), 4)

if __name__ == "__main__":
    # Unit tests: -ut
    unittest.main(argv=[" "])

File: kmp.cc
// kmp.cc - Kmp program by Sergey 2016

#include <bits/stdc++.h>
using namespace std;

///////////////////////////////////////////////////////////////////////////////
// Kmp (Main Program)
///////////////////////////////////////////////////////////////////////////////


vector<int> kmp_table(const string& W) {
    vector<int> T;
    assert(W.size() > 0);
    int cnd = -1;
    T.push_back(-1);
    for(auto c: W) {
        while (true) {
            if (cnd == -1 || W[cnd] == c) {
                cnd += 1;
                T.push_back(cnd);
                break;
            }
            cnd = T[cnd];
        }
    }
    return T;
}


int kmp_search(const string& W, const string& S) {
    int cnd = 0;
    auto T = kmp_table(W);
    for(unsigned int i=0; i < S.size(); i++) {
        while (true) {
            if (cnd == -1 || W[cnd] == S[i]) {
                cnd += 1;
                break;
            }
            cnd = T[cnd];
        }
        if (cnd == int(W.size())){
            return i - cnd + 1;
        }
    }
    return -1;
}



///////////////////////////////////////////////////////////////////////////////
// Unit tests
///////////////////////////////////////////////////////////////////////////////


// CPlusPlusUnit - C++ Unit testing TDD framework (github.com/cppunit/cppunit)
class Cppunit { public:
    #define CHECK(a,b)  check<long long>(a, b, #a, #b, __FILE__, __LINE__, __FUNCTION__);
    #define CHECKT(a)   check<bool>(a, true, #a, "true", __FILE__, __LINE__, __FUNCTION__);
    #define CHECKS(a,b) check<cs>(a, b, #a, #b, __FILE__, __LINE__, __FUNCTION__);
    typedef const std::string& cs;
    int checks, fails; std::ostringstream serr; std::istringstream *in;
    Cppunit() { checks = fails = 0;}
    void test_cin(cs s){ in = new std::istringstream(s); std::cin.rdbuf(in->rdbuf()); }
    void fail_hdr(cs stra, cs strb, cs file, int line, cs func) {
        serr << "==================================================" << std::endl;
        serr << "FAIL: " << func << std::endl;
        serr << "--------------------------------------------------" << std::endl;
        serr << "File \"" << file << "\", line " << line << " in " << func << std::endl;
        serr << "  Checking " << stra << " == " << strb << std::endl;
    }
    template <typename T> void check(T a, T b, cs stra, cs strb, cs file, int line, cs func) {
        checks++; if (a == b) { std::cout << "."; return; }
        fails++; std::cout << "F"; fail_hdr(stra, strb, file, line, func);
        serr << "  Error: \"" << a << "\" ! = \"" << b << "\"" << std::endl << std::endl;
    }
    virtual void single_test() {}
    virtual void test_list() { single_test(); }
    double dclock() { return double(clock()) / CLOCKS_PER_SEC; }
    int status() {
        std::cout << std::endl; if (fails) std::cout << serr.str();
        std::cout << "--------------------------------------------------" << std::endl;
        std::cout << "Ran " << checks << " checks in " << dclock() << "s" << std::endl << std::endl;
        if (fails) std::cout << "FAILED (failures=" << fails << ")"; else std::cout << "OK" << std::endl;
        return fails > 0;
    }
    int run() { std::streambuf* ocin = std::cin.rdbuf(); test_list(); std::cin.rdbuf(ocin); return status(); }
};


class MyCppunit: public Cppunit {

    void single_test() {

        // KMP table generation
        CHECK(kmp_table("A")[0],  -1);
        CHECK(kmp_table("AA")[2], 1);
        CHECK(kmp_table("AB")[2], 0);
        CHECK(kmp_table("ABAB")[4], 2);
        CHECK(kmp_table("ABCDABA")[7], 1)

        // KMP search
        CHECK(kmp_search("A", ""), -1);
        CHECK(kmp_search("AB", "B"), -1);
        CHECK(kmp_search("A", "BBBB"), -1);
        CHECK(kmp_search("A", "A"), 0);
        CHECK(kmp_search("BA", "BA"), 0);
        CHECK(kmp_search("A", "BA"), 1);
        CHECK(kmp_search("AS", "BDAS"), 2);
        CHECK(kmp_search("ABAC", "ABABABACB"), 4);

    }
};

int main() {
    // Unit tests: -ut
    return (new MyCppunit)->run();
}

